### ISSUE TITLE
Add methods for deleting nodes in model classes

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#752]
It should also delete the associated edges and parameters.

**Comment by kris-singh** [comment]
@ankurankan What do you mean deletion of parameters i did not add_parameters in Base.py file ?

**Comment by yashu-seth** [comment]
@kris-singh 
```python
>>> from pgmpy.models import BayesianModel
        >>> from pgmpy.factors.discrete.CPD import TabularCPD
        >>> student = BayesianModel([('diff', 'grades'), ('intel', 'grades')])
        >>> grades_cpd = TabularCPD('grades', 3, [[0.1,0.1,0.1,0.1,0.1,0.1],
        ...                                       [0.1,0.1,0.1,0.1,0.1,0.1],
        ...                                       [0.8,0.8,0.8,0.8,0.8,0.8]],
        ...                         evidence=['diff', 'intel'], evidence_card=[2, 3])

```

If you remove grades from the model it should also remove grades_cpd.

**Comment by ankurankan** [comment]
@kris-singh I have implemented this for the `BayesianModel`, you can have a look at it.

**Comment by kris-singh** [comment]
rather than implementing it for all model classes i was thinking that i would rather implemete the remove nodes method in the directed graph and undirected model.

**Comment by ankurankan** [comment]
@kris-singh I don't think that would be a great idea. Since the base Directed and Undirected graphs aren't supposed to deal with the parameters of the model. Also different models might have different parameterization, so might be a bit difficult to handle in the base classes.

**Comment by kris-singh** [comment]
@ankurankan @yashu-seth 
1. For factor graphs we could do something like this that if we remove a factor we just delete it, but if we remove a node we have recalculate the factor nodes its associated with(which basically is marginalizing that factor node wrt the deleted node)
2. For LinearGaussianBayesian Model we could call the super class delete method to deal with deletion.
p.s I am thinking of factors as TabularCPD so marginalisation would be easy.
3. i think for markov model it is very similar to the code in BayesianNetwork
what do you guys think ?

**Comment by raghavg7796** [comment]
@ankurankan @yashu-seth @kris-singh  For deleting factors in factor graph, there is already a method. And I think marginalize and delete should be two seperate methods. Deleting a node just deletes a node and all factors associated with it. While marginalize should marginalize with respect to a given variables. If you haven't worked a lot, I would like to implement the marginalize methods.

**Comment by yashu-seth** [comment]
@kris-singh The ideas seem feasible. If you want you can proceed with a PR.

**Comment by yashu-seth** [comment]
@raghavg7796 Deleting a node should not delete all the factors associated with it, but it should marginalize the factors for that variable node. And marginalize methods are only for factors not for models. So there is no need to implement marginalize methods for this. Please feel free to clarify if you meant something else.

**Comment by kris-singh** [comment]
@raghavg7796 there is no method for deleting a node in factor graph also, while you delete a node you dont delete its factors(unless the factors has no incoming edges from the nodes).  And while deleting a node you need to marginalize factors. For the present i think its best i implement this as having 2 people devlop the same thing would require a lot of co-ordination from both our ends. 
@yashu-seth Yes i will start working on these ideas

**Comment by kris-singh** [comment]
@ankurankan @yashu-seth remove_factor method is wrongly implemented 
```python
>>> from pgmpy.models import FactorGraph
>>> from pgmpy.factors.discrete import DiscreteFactor
>>> G = FactorGraph()
>>> G.add_nodes_from(['a', 'b', 'c'])
>>> phi1 = DiscreteFactor(['a', 'b'], [2, 2], np.random.rand(4))
>>> G.add_edge('a',phi1)
>>> G.add_factors(phi1)
>>> [G.remove_factors(phi1)]
```
for this `phi1` is removed from `G.factors` list but not from `G.nodes()` list. So it should be deleted from both places
I have remove_factor to
```python
for factor in factors:
    self.factors.remove(factor)
    super(FactorGraph,self).remove_node(factor)
```

**Comment by yashu-seth** [comment]
@kris-singh The users are suppose to add factors separately and the ```check_model``` method ensures if factors are defined for each factor node or not. So I don't think this behavior is required. I think the only thing to be ensured should be that when you delete a node, the factors should appropriately be marginalized and when you delete a factor node that factor should be removed.
@ankurankan What do you think?

**Comment by kris-singh** [comment]
@yashu-seth i am having a hard time getting this. Should the user call check_model after every operation.
That doesn't seem a good idea. After deletion of a factor it still present in the node of the graph that seems erroneous to me.
@ankurankan My PR is ready for the factors graph node_removal but it uses the change in remove_factors can you please suggest what to do?

**Comment by kris-singh** [comment]
@yashu-seth @ankurankan 
```python
>>>import numpy as np
>>>from pgmpy.models import MarkovModel
>>>from pgmpy.factors.discrete import DiscreteFactor
>>>student = MarkovModel([('Alice', 'Bob'), ('Bob', 'Charles')])
>>>factor = DiscreteFactor(['Alice', 'Bob'], cardinality=[2, 2],values=np.random.rand(4))
>>>student.add_edge('Alice',factor)
>>>student.add_factors(factor)
>>>student.remove_factors(factor)
>>>print student.edges()
>>>print student.check_model()
```
output:
```python
[(<DiscreteFactor representing phi(Alice:2, Bob:2) at 0x7fcc7985e0d0>, 'Alice'), ('Bob', 'Charles'), ('Bob', 'Alice')]
True
```
Though we delete the factor the edge is not deleted between factor and node. This is consistent across all the remove_factor function in all the models

**Comment by ankurankan** [comment]
@kris-singh @yashu-seth I don't get where the whole conversation is going. Can someone give me a final gist ? Also @kris-singh please have a look at github markdown and try to use that to properly format your code.

**Comment by kris-singh** [comment]
@ankurankan this about remove_factor method in the models. I was of the view that remove_factor should not only remove the factors from the factors list but also from the nodes list. As i have illustrated in the code it  that is not done the edges between the deleted factor and node will still remain in the model. I have also explained this in the PR if you can have look.

**Comment by yashu-seth** [comment]
@ankurankan When we delete a node we need to delete/marginalize the factors for that node too. In factor graph, nodes can be factors. So when we delete a factor node, we should delete that factor. But if we remove a factor from a factor graph, should we also remove the corresponding factor node as well? This behavior will not be consistent with other models but now I think this does make sense for factor graphs.

**Comment by kris-singh** [comment]
@yashu-seth not only for factor graphs but it should be done for every model that has the remove_factor method as my previous code clearly shows that in the markov model this would create a spurious edge which is also not detected by check_model method.

**Comment by yashu-seth** [comment]
@kris-singh Can you please reproduce that markov model example you are talking about.

**Comment by kris-singh** [comment]
@ankurankan @yashu-seth please look at the following gist
https://gist.github.com/kris-singh/5b4fdfef4bb0bd589d255ae2d403a14d

**Comment by ankurankan** [comment]
@yashu-seth @kris-singh I don't think we need the `remove_factor` method at all. There should be just one method `remove_node` and it should delete the nodes as well as the factors from the model. 

And we will need to refactor a few things according to this. For example we won't be needing the add_factor method

**Comment by yashu-seth** [comment]
@ankurankan How would we incorporate factors with models then?

**Comment by ankurankan** [comment]
@yashu-seth Since in this case we really don't have a separate parameterization of the model. So, in simple terms we are just trying to create a model structure. And therefore it makes sense to remove the parts that deal specifically with model parameterization and only work with the model assuming that we are dealing with just the structure.

**Comment by yashu-seth** [comment]
@ankurankan Are you talking about ```FactorGraph``` only?

**Comment by ankurankan** [comment]
@yashu-seth Yes, I am sorry I wasn't clear before.

**Comment by yashu-seth** [comment]
@ankurankan Okay. @kris-singh suggested [here](https://gist.github.com/kris-singh/5b4fdfef4bb0bd589d255ae2d403a14d) that there could be cases where we have a factor node in markov models. I don't think we need to change the remove_factor method for such cases because we have ```FactorGraph```. What do you think?

**Comment by ankurankan** [comment]
@yashu-seth Why would we have a factor node in a Markov Model ?

**Comment by yashu-seth** [comment]
@ankurankan @kris-singh There would not be. That is why I am also not in the favor of having such behavior.

**Comment by kris-singh** [comment]
@yashu-seth @ankurankan then we would have to refactor the markov model and other classes that have add_factor method in them.

**Comment by yashu-seth** [comment]
@kris-singh No, you have to do it only for ```FactorGraph```.

**Comment by kris-singh** [comment]
@yashu-seth i mean to say that we would have to remove add_factor method in markov model if it dosen't make sense to have it as method if you are never going to use it.

**Comment by yashu-seth** [comment]
@kris-singh We will use those methods in MarkovModel. Just not in FactorGraph. We will not have a factor node in markov model but we will have factors in it. remove_factor method removes the factors associated with a model.

**Comment by kris-singh** [comment]
@yashu-seth okay then i think a check needs to be added saying that edge between factor node and edge node is not allowed. This at present is not there. @ankurankan if this issue is resolved then i think you can review #804 as i had followed the same technique there that we have agreed upon.

**Comment by yashu-seth** [comment]
@kris-singh I don't think that such a check is particularly required. Its not that such edges are incorrect its just that a user would not create such edges if dealing with models like the Markov model.

**Comment by kris-singh** [comment]
@yashu-seth if i understand this clearly we now will have no add_factor method and remove_factor method in FactorGraph. Instead, we will have just add_nodes. So for factor graph should we now create a new node and define a CPD on it (basically treat it as a factor). I am not in favor having no add_factor method in FactorGraph. As most of the function are dependent on the factors being present. Maybe we could do somthing with remove_factor but i still strongly think that add_factor method should be there.

**Comment by kris-singh** [comment]
I think i can push the functionality of remove_factor inside of remove_node. Checking if the passed argument is a factor or a node.

**Comment by yashu-seth** [comment]
@kris-singh For FactorGraph we can define factors and add it as nodes. 
I don't think removing add_factor would be difficult. I don't find functions that are dependent on FactorGraph.add_factors. Could you please elaborate?

**Comment by kris-singh** [comment]
@yashu-seth get_partition_function, to_markov_model, get_variable_nodes etc all use self.factors in one way or another. If we do away with factor nodes then all of these would have to refactor. But yes i think the above strategy that you suggest of adding factors as nodes would work, but we would have to keep a maintain a separate list for factors

**Comment by ankurankan** [comment]
@kris-singh Yes we will need to maintain a list because our inference methods access those.

**Comment by kris-singh** [comment]
@ankurankan @yashu-seth just one more question should i overload the add_node function from the undirected graph for adding the factors to the factor graph as we need to check that factors should be defined on the variables in the model. I was thinking of not deleting the add_factor method otherwise, we would have to rewrite a lot of tests. I have removed the remove_factor method. And changed the tests accordingly.

**Comment by kris-singh** [comment]
@ankurankan @yashu-seth can you look at this gist and tell me what the problem is.
https://gist.github.com/kris-singh/8b89ae2ea40c6a966795556e788212bd
Did you get the problem

**Comment by kris-singh** [comment]
Can anyone expalin the above problem.

**Comment by ankurankan** [comment]
@kris-singh Sorry, for the late reply. I am not able to figure out the issue. Can you post this on stackoverflow ? And please post the question's link here.

**Comment by kris-singh** [comment]
@ankurankan okay i will do that now

**Comment by kris-singh** [comment]
@ankurankan networkx/networkx#2349
i think we need to look at the hash function

**Comment by ankurankan** [comment]
@kris-singh Try posting it to stackoverflow. And don't post the whole code. Just the relevant parts.

**Comment by kris-singh** [comment]
@ankurankan https://stackoverflow.com/questions/41785811/networkx-hash-function-for-objects
Also there are not test for checking the hash and eq function in the test_factor_graph. I think we should look into it
